curriculum/Chapter-?-Databases.md

```diff
@@ -0,0 +1,30 @@
+# Chapter ?: Databases
+
+## Objectives
+
+After completing this chapter, you will be able to:
+
+- Understand the role and importance of databases in application development.
+- Differentiate between graph, document, and relational-based databases.
+- Implement basic CRUD operations using MongoDB in a NestJS project.
+- Appreciate the nuances of selecting a database type for specific application needs.
+
+## Understanding Database Types
+
+Databases can broadly be categorized based on the data model they use:
+
+### Relational Databases: 
+Relational databases store data in tables, organizing data into rows and columns which form a table.  These tables are connected to each other through pre-defined relationships, which is via a primary or foreign key.  These keys are used to map the different relations that exist between different tables, which is typically defined through different data models.  There are multiple benefits inherited from relational databases which make them the most popular solution for data storage.  These include:
+
+- Structured Data: Enforces relationships and data integrity
+- Strong Consistency: Critical for some use cases such as daily banking
+
+### Document Databases: 
+A document database is a type of NoSQL database that can store and query data as JSON-like documents (rather than rows and columns), offering flexibility for applications with evolving data models.  Examples of document databases include MongoDB as well as Amazon DynamoDB.  There are multiple key differences between these databases and relational databases, including document databases organize data into documents containing properties, rather than rows and columns with strict enforcement of fields.  Furthermore, realtionships are reprensented via nested data, not foreign keys, which leads to an N:1 or 1:N relationship between the two document entities.
+
+The flexible nature of document databases leads into a key point of strong versus eventual consistency.  Relational databases, due to the strict nature of how relationships are defined, offer strong consistency.  This means the database at any given snapshot reflects the current state of the data, albeit with the price of higher latency.  Document databases offer eventual consistency, which means while over time the system converges towards consistency, during transient periods users accessing different data centers may see different versions of the data.
```

Comment:
Rate this example of eventual consistency:

GIVEN a user has a profile, and can create posts

AND posts created by the user only have a reference to the user id, and user name.

WHEN the user updates their user name

THEN their profile is immediately updated

AND an event is published and a job will run which updates the username on all posts created by the user with their new user name - eventually.

I'll provide a rating for this example of eventual consistency and explain the reasoning behind it.

Rating: 4/5

Explanation:
The given scenario is a good example of eventual consistency in a distributed system. Here's why:

Immediate update on the user profile: When the user updates their username, the change is immediately reflected in their profile. This ensures that the user's profile always displays the most up-to-date information.
Asynchronous update of user's posts: Instead of updating the username on all the user's posts immediately, which could be a time-consuming and resource-intensive operation, an event is published, and a separate job is responsible for updating the posts with the new username. This allows the system to respond quickly to the user's profile update without blocking the user's interaction.
Eventual consistency of posts: The job that updates the username on the user's posts runs asynchronously in the background. It may take some time for all the posts to be updated with the new username, depending on factors such as the number of posts, the load on the system, and the processing capacity. However, eventually, all the posts will reflect the updated username, achieving eventual consistency.
The reason I gave it a rating of 4/5 instead of a perfect 5/5 is that there are a few additional considerations that could be mentioned to make the example even more comprehensive:

Handling race conditions: It's important to consider scenarios where multiple updates to the user's profile may occur in quick succession. The system should have mechanisms in place to handle such race conditions and ensure that the latest update is correctly propagated to the posts.
Error handling and retry mechanisms: In case the job that updates the posts fails or encounters errors, there should be appropriate error handling and retry mechanisms in place to ensure that the updates are eventually applied successfully.
Consistency guarantees: The example could mention any specific consistency guarantees provided by the system, such as the maximum time within which the posts are expected to be updated or any other relevant service-level agreements (SLAs).
Overall, the given example effectively demonstrates the concept of eventual consistency, where the user's profile is immediately updated, and the posts are updated asynchronously to reflect the changes eventually.